irmaseal-core/src/stream/opener.rs

```diff
@@ -1,143 +1,134 @@
-use crate::stream::util::ArchiveReader;
 use crate::stream::*;
 use crate::*;
 
-use core::convert::TryInto;
 use arrayref::array_ref;
-use ctr::stream_cipher::{NewStreamCipher, StreamCipher};
+use arrayvec::ArrayVec;
+use core::convert::TryFrom;
+use core::convert::TryInto;
 use hmac::Mac;
 
-/// First stage opener of an IRMAseal encrypted bytestream.
+/// Opener of an IRMAseal encrypted bytestream.
 /// It reads the IRMAseal header, and yields the recipient Identity for which the content is intended.
 ///
 /// Enables the library user to lookup the UserSecretKey corresponding to this Identity before continuing.
-pub struct OpenerSealed<R: Readable> {
-    ar: ArchiveReader<R, [u8; MAX_METADATA_SIZE]>,
-}
-
-/// Second stage opener of an IRMAseal encrypted bytestream.
-///
-/// **Warning**: will only validate the authenticity of the plaintext when calling `validate`.
-pub struct OpenerUnsealed<R: Readable> {
-    aes: SymCrypt,
-    hmac: Verifier,
-    r: R,
-    resultbuf: Option<[u8; BLOCKSIZE]>,
+pub struct OpenerSealed<R: AsyncRead + Unpin> {
+    input_reader: R,
+    metadata_buf: ArrayVec<[u8; MAX_METADATA_SIZE]>,
 }
 
-impl<R: Readable> OpenerSealed<R> {
+impl<R: AsyncRead + Unpin> OpenerSealed<R> {
```

Comment:
Opening an irmaseal stream has two stages: 
 1. reading the metadata from the input stream and returning the identity
 2. receive the private key of the identity, decrypt the encrypted content of the input stream and stream this to the output.

I now do this by two methods, the `new` constructor for stage 1 and the `unseal` method for stage 2. The input stream is captured in the `OpenerSealed` object since it contains both the metadata and the content and we therefore need it in both stages.

Disadvantage of this approach is that we offer an `unseal` method which can only be called once and this is not explicitly enforced by the compiler. This can maybe be confusing.

A solution would be to let `OpenerSealed` implement `FnOnce` such that `new` returns a function that can be executed once. The only disadvantage of this approach is that there is no async variant of `FnOnce`, so that is a bit challenging. Then I have to walk along the borders of Rust again in the hope I get it working.

Do you have ideas on this?

Reply:
I overlooked the fact that `unseal` already uses a `mut seal`, so it consumes the `OpenerSealed` object when it is called, so it cannot be called multiple times.